content/blog/2019/nov/grpc-a-powerful-way-to-improve-golang-apis/index.md

```diff
@@ -0,0 +1,286 @@
+---
+title: 'GRPC: A powerful way to improve your Golang APIs'
+date: '2019-11-17T08:00:00Z'
+cover: './cover.jpg'
+---
+
+![Photo by Israel Palacio on Unsplash](cover.jpg)
+
+Web APIs are everywhere, with REST been one of the most popular ways to distribute it. With recent technologies, there are improved ways to implement it, GRPC been one of them.
+
+# Why REST is popular and what are its pitfalls?
+
+Companies used to write web-services in SOAP until REST got enough hype to be the next big thing, although the concept been around since 2000. It was like a fresh breeze for developers. It had a low learning curve, simplified payload (usually JSON instead of XML), schema-less approach and no need for using specific clients/server generators.
+
+At a certain point, some of its strengths shown to be some of its weaknesses. Not having a schema initially seemed a good idea, but developers eventually realised it is required. It can be useful for validation, documentation, contract agreement between teams and code-generation, to mention a few. Initiatives such as Swagger, RAML and most recently OpenAPI popped up trying to fill these gaps.
+
+Even though, a lack of an official standard to guide through decision processes usually ended up in analysis paralysis in big projects. Besides, code generation from most of these tools was not as good and ready to be used (example: swagger-tools).
+
+JSON is the most preferred way to return data from REST APIs. It is an excellent choice for public services used by front-end applications and third-parties. Due to its overhead on transporting and decoding, it is not well optimised for internal services communication.
```

Comment:
JSON is the preferred way ... optimised for communication between internal services.